lib/puppet/transaction/report.rb

```diff
@@ -386,6 +394,101 @@ def self.default_format
 
   private
 
+  # Using the last applied values from our persistence store, and the current
+  # report data, iterate across all events finding any corrective_change that
+  # might exist and marking the event, resource and report appropriately.
+  def process_resource_inferences
+    persistence = Puppet::Transaction::Persistence.new
+    persistence.load
+
+    new_resources = {}
+
+    resource_statuses.each do |name, status|
+      # Initialize data structure for resource
+      new_resources[name] ||= {}
+      params = new_resources[name]["parameters"] ||= {}
+
+      # Populate new hash with events mapped to params
+      param_to_event = {}
+      status.events.each do |ev|
+        param_to_event[ev.property] = ev
+      end
+
+      status.real_resource.parameters.each do |pname,param|
+        # Skip if if the data is sensitive, or if its just a property
+        next if !param.is_a?(Puppet::Property) || param.sensitive
+
+        # Create new resource parameter definition in new hash
+        new_param_values = params[pname.to_s] ||= {}
+        event = param_to_event[pname.to_s]
+
+        # Retrieve the old system value
+        old_system_value = persistence.get_system_value(name, pname.to_s)
+
+        # Store the new system value for persistence
+        new_param_values["system_value"] = new_system_value(param, event, old_system_value)
+
+        # Calculate corrective change if there is a related event
+        event.corrective_change = is_corrective_change(param, event, old_system_value) if event
+      end
+
+      # Mark resource and report with corrective_change if there is one
+      status.events.each do |e|
+        if e.corrective_change
+          status.corrective_change = true
+          @corrective_change = true
+          break
+        end
+      end
+    end
+
+    # Persist our new values to the resources storage.
+    persistence.resources = new_resources
+    persistence.save
+  end
+
+  # Given an event and its property, calculate the system_value to persist
+  # for future calculations.
+  # @param [Puppet::Transaction::Event] event event to use for processing
+  # @param [Puppet::Property] property correlating property
+  # @param [Object] system_value system_value from last transaction
+  # @return [Object] system_value to be used for next transaction
+  def new_system_value(property, event, system_value)
```

Comment:
Well, in this particular case it is actually required for all non-sensitive properties, not just ones that generate an event. Event might not be the place, is my point, probably ResourceHarness or Property.